File: src/com/iwf/demo/subscription/WaitForPeriodState.java
package com.iwf.demo.subscription;

import com.iwf.StateDecision;
import com.iwf.StateMovement;
import com.iwf.WorkflowState;
import com.iwf.attributes.QueryAttributesRO;
import com.iwf.attributes.QueryAttributesRW;
import com.iwf.attributes.SearchAttributesRO;
import com.iwf.attributes.SearchAttributesRW;
import com.iwf.command.CommandRequest;
import com.iwf.command.CommandResults;
import com.iwf.command.TimerCommand;
import com.iwf.demo.subscription.models.Customer;

import java.util.ArrayList;

import static com.iwf.demo.subscription.ChargeCurrentPeriodState.WF_STATE_CHARGE_CURRENT_PERIOD;
import static com.iwf.demo.subscription.SubscriptionOverState.WF_STATE_SUBSCRIPTION_OVER;
import static com.iwf.demo.subscription.WelcomeEmailState.WF_STATE_SEND_WELCOME_EMAIL;

class WaitForPeriodState implements WorkflowState<Void> {

    public static final String WF_STATE_WAIT_FOR_NEXT_PERIOD = "waitForNextPeriod";

    @Override
    public String getStateId() {
        return WF_STATE_SEND_WELCOME_EMAIL;
    }

    @Override
    public Class<Void> getInputType() {
        return Void.class;
    }

    @Override
    public CommandRequest execute(final Void nothing, final SearchAttributesRO searchAttributes, final QueryAttributesRO queryAttributes) {
        final Customer customer = queryAttributes.get(SubscriptionWorkflow.QUERY_ATTRIBUTE_CUSTOMER);

        return CommandRequest.forAllCommandCompleted(
                new TimerCommand((int) (System.currentTimeMillis() / 1000) + customer.getSubscription().getPeriodsInSubscription())
        );
    }

    @Override
    public StateDecision decide(final Void nothing, final CommandResults commandResults,
                                final SearchAttributesRW searchAttributes, final QueryAttributesRW queryAttributes) {
        ArrayList<StateMovement> nextStates = new ArrayList();
        final Customer customer = queryAttributes.get(SubscriptionWorkflow.QUERY_ATTRIBUTE_CUSTOMER);
        int currentPeriodNum = queryAttributes.get(SubscriptionWorkflow.QUERY_ATTRIBUTE_BILLING_PERIOD_NUMBER);
        if (currentPeriodNum < customer.getSubscription().getPeriodsInSubscription()) {
            queryAttributes.upsert(SubscriptionWorkflow.QUERY_ATTRIBUTE_BILLING_PERIOD_NUMBER, currentPeriodNum + 1);
            nextStates.add(new StateMovement(WF_STATE_CHARGE_CURRENT_PERIOD));
        } else {
            nextStates.add(new StateMovement(WF_STATE_SUBSCRIPTION_OVER));
        }

        return new StateDecision(
                nextStates
        );
    }
}

File: gosdk/demo/subscription/waitForPeriodState.go
package subscription

import (
	"fmt"
	"time"

	"github.com/longquanzheng/iwf/gosdk/iwf"
)

type WaitForPeriodState struct {

}
const WF_STATE_WAIT_FOR_NEXT_PERIOD = "waitForNextPeriod"

func (w WaitForPeriodState) GetStateId() string {
	return WF_STATE_WAIT_FOR_NEXT_PERIOD
}

func (w WaitForPeriodState) GetInputType() iwf.NewTypePtr {
	return nil
}

func (w WaitForPeriodState) GetSearchAttributesLoadingPolicy() iwf.AttributeLoadingPolicy {
	return nil
}

func (w WaitForPeriodState) GetQueryAttributesLoadingPolicy() iwf.AttributeLoadingPolicy {
	return nil
}

func (w WaitForPeriodState) Execute(ctx iwf.WorkflowContext, input interface{}, searchAttributes iwf.SearchAttributesRO, queryAttributes iwf.QueryAttributesRO) (iwf.CommandRequest, error){
	customer, ok := input.(*Customer)
	if !ok{
		return nil, fmt.Errorf("cannot get customer from input")
	}
	waitTime := time.Duration(customer.Subscription.BillingPeriodInSeconds) * time.Second
	return iwf.WaitForAllCommandsCompleted(
		iwf.NewTimerCommand(time.Now().Add(waitTime)),
	), nil
}

func (w WaitForPeriodState) Decide(ctx iwf.WorkflowContext, input interface{}, commandResults iwf.CommandResults, searchAttributes iwf.SearchAttributesRW, queryAttributes iwf.QueryAttributesRW) (iwf.StateDecision, error) {
	panic("implement me")
}

var _ iwf.WorkflowState = (*WaitForPeriodState)(nil)